Use rxjs 5 WebSocketSubject

deontologician opened the pull request:
Implements #276, #527 (client side at least), and should fix #595 directly.

Turns out it can't be used in a "composed" way since it does some stuff that needs to be overridden. So this is subclassing WebSocketSubject in a way that kind of throws out a bit of its functionality and implements some functionality we need like handshakes and messages that complete requests (`state: complete` should finish a multiplexed observable)

This is also laying the foundation for #358 by keeping an internal map of requests so they can be re-sent later upon reconnection. That functionality won't land in 2.0 however

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/rethinkdb/horizon/619)

<!-- Reviewable:end -->

marshall007 commented on client/src/socket.js:
```diff
@@ -31,14 +32,164 @@ class ProtocolError extends Error {
+    this.handshakeSub.add(keepAliveSub)
+  }
+
+  multiplex(rawRequest) {
```
@deontologician I haven't tested any of this, but I think you might be able to re-use most of the `multiplex` logic from the base class like so:

``` js
multiplex(rawRequest) {
  const requestId = this.requestCounter++

  const subMsg = () => {
    if (this.activeRequests.size === 0) {
      // send handshake only if we're the first request to be
      // subscribed to
      this.sendHandshake()
    }

    const request = Object.assign({ request_id: requestId }, rawRequest)
    this.activeRequests.set(requestId, { request })
    return request
  }

  const unsubMsg = () => {
    this.deactivateRequest(requestId)
    return { request_id: requestId, type: 'end_subscription' }
  }

  return super.multiplex(subMsg, unsubMsg, msg => msg.request_id === requestId)
}
```

deontologician replied:
This is true, but once I trimmed out the generality I didn't need (arbitrary filter function, tryCatch around stuff I know can't throw an exception) the function was about the same size, so I just reimplemented it and eliminated some extra function calls and closures

marshall007 replied:
They are intentionally wrapping the `tryCatch` in a closure because that's a fairly well-known workaround for ensuring the calling function can still be optimized by V8, meaning it has minimal performance impact at runtime (`bluebird` uses the same trick). I would imagine bytes saved + their fancy compiler tricks is actually better in terms of performance than reducing closures here.

The real advantage is that it's easier to read and thus make changes. I think we probably can/should do the same thing in `sendHandshake()` as well.

marshall007 replied:
https://github.com/petkaantonov/bluebird/wiki/Optimization-killers

deontologician replied:
Ok, I am convinced! I will keep my hand out of optimization efforts. It's probably best to re-use as much of the underlying class anyway

marshall007 commented on client/src/socket.js:
```diff
@@ -136,84 +138,77 @@ class HorizonSocket extends WebSocketSubject {
-      // Now we subscribe to messages from the server with our
-      // requestId.
-      const subscription = this.subscribe({
+      this.handshake.ignoreElements().concat(
```
@deontologician might be able to avoid creating the inner observer (`subscriber`) here:

``` js
return this.handshake.ignoreElements().concat(
  super.multiplex(subMsg, unsubMsg, res => res.request_id === requestId)
)
.takeUntil(res => res.state === 'complete')
.map(res => {
  if (res.error !== undefined) {
    // This is an error just for this request, not the
    // entire connection
    throw new ProtocolError(res.error, res.error_code)
  }
  if (res.data !== undefined) {
    // Forward on the response if there's a data field
    return res
  }
  if (res.state === 'synced') {
    // Create a little dummy object for sync notifications
    return { type: 'state', state: 'synced' }
  }
})
```

In fact, the body of the `.map()` could be moved to `resultSelector`. `throw`ing from there will propagate as an error through the observable chains just fine. I think doing so would also allow you to remove that same logic from the `handshakeSub` since the responses coming out of `super.multiplex` would already be formatted or thrown as errors.

deontologician replied:
Ok, I think I can incorporate this

deontologician replied:
Ended up needing to emit an extra event on `complete`, since takeWhile doesn't return the matching event

deontologician replied:
also `concatMap`, since sometimes we need to emit multiple events for a single message from the server

marshall007 replied:
👍 looks really good!

deontologician commented:
So, blargle. The subMsg wants a synchronous result, and we need to wait for the handshake to complete before emitting the subscription. And it _must_ have a subscription message, multiplex will send `null` or `undefined` over the socket if you refuse to set one (which for us will be a protocol error). It would be ideal if subMsg and unsubMsg could be provided by observables, since then we could wait for the handshake. As it is the only option is reimplementing multiplex

deontologician commented:
Ok, actually nevermind, it looks like concating the multiplex observable onto the handshake suffices to prevent an early request

deontologician commented:
This is still pending some fixes that are in the master branch of rxjs5 but haven't been merged yet (specifically https://github.com/ReactiveX/rxjs/pull/1790)

deontologician commented:
Ok, most recent things I've run into:
1. [x] horizon-core.js doesn't create a Horizon global
2. [x] Handshake seems to be firing off more than once per websocket connection.
3. [x] Tests still failing, though this may be a result of 2

The "It works!" page test is going though. So this is in the final stretch

alf commented:
The fixes in rxjs seems to be included in beta10 which is out.

deontologician commented:
All tests are now passing. Going to clean up the packaging situation now (hopefully simple...)

deontologician commented:
Also thanks to @marshall007 for a ton of help, this has been unexpectedly frustrating

Commit:
refactor: Socket uses WebSocketSubject

- new methods hzRequest and makeRequest on socket
- keepalives added (resolves #527)

marshall007 commented:
@deontologician congrats! No problem! I think we learned a lot in the process... although, if asked, I'm still not sure I could tell you anything about how `WebSocketSubject` is supposed to work, haha.

deontologician commented:
Ok, I think this is ready to go (assuming circleci doesn't complain). I'm going to do the build system changes in another PR since this one is already pretty gnarly.

deontologician commented:
@dalanmiller or @Tryneus interested in reviewing?

deontologician commented:
Ok! It turns out I'm not crazy, I just forgot I needed to fix the underlying WebSocketSubject this whole this is based on. There's a PR I opened in upstream: https://github.com/ReactiveX/rxjs/pull/1831

Once that's merged and released, we can bump the rxjs version on this library and merge this PR

Commit:
fix: hack _subscribe

Needed until https://github.com/ReactiveX/rxjs/pull/1831 is merged

deontologician commented:
Since ReactiveX/rxjs#1831 may take some time to get merged, I added a hack to override `_subscribe` in the meantime. This now passes all tests and should be good to merge.
